presto-main/src/main/java/com/facebook/presto/operator/LookupJoinOperator.java

```diff
@@ -134,94 +194,377 @@ public boolean needsInput()
+            }
+            else {
+                Page remaining = pageTail(currentPage, currentPosition);
+                restoreProbe(remaining, currentJoinPosition, currentProbePositionProducedRow, joinSourcePositions, spillInfoSnapshot);
+            }
+        }
     }
 
-    private Page getOutput(LookupSource lookupSource)
+    private void processProbe(LookupSource lookupSource)
     {
-        // join probe page with the lookup source
+        verify(probe != null);
+
         DriverYieldSignal yieldSignal = operatorContext.getDriverContext().getYieldSignal();
-        if (probe != null) {
-            while (!yieldSignal.isSet()) {
-                if (probe.getPosition() >= 0) {
-                    if (!joinCurrentPosition(lookupSource, yieldSignal)) {
+        while (!yieldSignal.isSet()) {
+            if (probe.getPosition() >= 0) {
+                if (!joinCurrentPosition(lookupSource, yieldSignal)) {
+                    break;
+                }
+                if (!currentProbePositionProducedRow) {
+                    currentProbePositionProducedRow = true;
+                    if (!outerJoinCurrentPosition(lookupSource)) {
                         break;
                     }
-                    if (!currentProbePositionProducedRow) {
-                        currentProbePositionProducedRow = true;
-                        if (!outerJoinCurrentPosition(lookupSource)) {
-                            break;
-                        }
-                    }
-                }
-                currentProbePositionProducedRow = false;
-                if (!advanceProbePosition(lookupSource)) {
-                    break;
                 }
-                statisticsCounter.recordProbe(joinSourcePositions);
-                joinSourcePositions = 0;
             }
+            currentProbePositionProducedRow = false;
+            if (!advanceProbePosition(lookupSource)) {
+                break;
+            }
+            statisticsCounter.recordProbe(joinSourcePositions);
+            joinSourcePositions = 0;
         }
+    }
+
+    private void restoreProbe(Page probePage, long joinPosition, boolean currentProbePositionProducedRow, int joinSourcePositions, SpillInfoSnapshot spillInfoSnapshot)
+    {
+        verify(probe == null);
+
+        addInput(probePage, spillInfoSnapshot);
+        verify(probe.advanceNextPosition());
+        this.joinPosition = joinPosition;
+        this.currentProbePositionProducedRow = currentProbePositionProducedRow;
+        this.joinSourcePositions = joinSourcePositions;
+    }
+
+    private Page pageTail(Page currentPage, int startAtPosition)
+    {
+        verify(currentPage.getPositionCount() - startAtPosition >= 0);
+
+        int[] retainedPositions = new int[currentPage.getPositionCount() - startAtPosition];
+        Arrays.setAll(retainedPositions, i -> startAtPosition + i);
+        return mask(currentPage, retainedPositions);
```

Comment:
Why are you masking instead of doing `currentPage.getRegion(startAtPosition, currentPage.getPositionCount())`.  If there is a good reason, add a comment

Reply:
I didn't know this method.